File: samples/java/Bingo/Server/Session/src/main/java/systems/zlink/samples/bingo/server/session/SessionServerApplication.java
package systems.zlink.samples.bingo.server.session;

import io.micrometer.core.instrument.MeterRegistry;

import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;

import systems.zlink.framework.codecs.protobuf.ZLinkProtobufCodec;
import systems.zlink.framework.configuration.ZLinkMeshNodeBuilder;
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode;
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;
import systems.zlink.framework.spring.EnableZLinkFramework;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;
import systems.zlink.samples.bingo.server.configuration.BingoMetricsReporter;
import systems.zlink.samples.bingo.server.configuration.BingoReadinessReporter;
import systems.zlink.samples.bingo.server.configuration.SampleApplication;
import systems.zlink.samples.bingo.server.configuration.SampleLocationStore;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTopology;
import systems.zlink.samples.bingo.server.session.sessions.BingoSession;

@EnableZLinkFramework
@EnableConfigurationProperties(SampleTopology.class)
@SpringBootApplication(
        proxyBeanMethods = false,
        scanBasePackageClasses = SessionServerApplication.class)
public final class SessionServerApplication {
    private SessionServerApplication() {}

    public static AutoCloseable run(String configPath) {
        return SampleApplication.start(SessionServerApplication.class, configPath)::close;
    }

    @Bean
    ZLinkFrameworkConfigurer sessionFramework(SampleTopology topology) {
        return options -> {
            options.addHandlersFromPackageOf(SessionServerApplication.class);
            options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL);
            options.codecs().use(ZLinkProtobufCodec.defaultCodec());
            options.configureLocations();
            // --8<-- [start:doc-bingo-session-register]
            ZLinkMeshNodeBuilder node = options.addRouteMesh(SampleNames.Mesh);
            node.listen(topology.selectedSessionRouterEndpoint()).setRoutingIdPrefix("session");
            node.objects().client();
            options.addClientServerChannel(SampleNames.ApiChannel).client();
            node.channelName(SampleNames.RoomSpotDiscovery).client();
            options.addStreamNode(SampleNames.StreamNode)
                    .bind(topology.selectedStreamEndpoint())
                    .enableActorDispatch()
                    .registerSession(BingoSession.class);
            // --8<-- [end:doc-bingo-session-register]
        };
    }

    @Bean
    ZLinkRedisLocationStore locationStore(SampleTopology topology) {
        return SampleLocationStore.create(topology);
    }

    @Bean(destroyMethod = "close")
    BingoMetricsReporter bingoMetricsReporter(MeterRegistry registry) {
        return new BingoMetricsReporter(registry, "session");
    }

    @Bean(destroyMethod = "close")
    BingoReadinessReporter bingoReadinessReporter(
            SampleTopology topology, ZLinkRouteMeshRuntime meshes) {
        return BingoReadinessReporter.session(topology, meshes);
    }
}

File: samples/kotlin/Bingo/Server/Session/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/session/SessionServerApplication.kt
package systems.zlink.samples.kotlin.bingo.server.session

import io.micrometer.core.instrument.MeterRegistry
import java.nio.file.Path
import kotlinx.coroutines.Dispatchers
import org.springframework.boot.WebApplicationType
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.core.env.StandardEnvironment
import systems.zlink.framework.codecs.protobuf.ZLinkProtobufCodec
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode
import systems.zlink.framework.kotlin.configureDispatch
import systems.zlink.framework.kotlin.useCoroutineHandlers
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime
import systems.zlink.framework.spring.EnableZLinkFramework
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer
import systems.zlink.samples.kotlin.bingo.server.configuration.BingoMetricsReporter
import systems.zlink.samples.kotlin.bingo.server.configuration.BingoReadinessReporter
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleLocationStore
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTopology
import systems.zlink.samples.kotlin.bingo.server.session.sessions.BingoSession

@EnableZLinkFramework
@EnableConfigurationProperties(SampleTopology::class)
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackageClasses = [SessionServerApplication::class],
)
class SessionServerApplication {
    @Bean
    fun sessionFramework(topology: SampleTopology): ZLinkFrameworkConfigurer =
        ZLinkFrameworkConfigurer { options ->
            options.addHandlersFromPackageOf(SessionServerApplication::class.java)
            options.useCoroutineHandlers(Dispatchers.Default)
            options.configureDispatch { messageFlow(ZLinkMessageFlowLogMode.NORMAL) }
            options.codecs().use(ZLinkProtobufCodec.defaultCodec())
            options.configureLocations()
            // --8<-- [start:doc-bingo-session-register]
            val node = options.addRouteMesh(SampleNames.Mesh)

            node.listen(topology.selectedSessionRouterEndpoint()).setRoutingIdPrefix("session")
            node.objects().client()
            options.addClientServerChannel(SampleNames.ApiChannel).client()
            node.channelName(SampleNames.RoomSpotDiscovery).client()
            options
                .addStreamNode(SampleNames.StreamNode)
                .bind(topology.selectedStreamEndpoint())
                .enableActorDispatch()
                .registerSession(BingoSession::class.java)
            // --8<-- [end:doc-bingo-session-register]
        }

    @Bean
    fun locationStore(topology: SampleTopology): ZLinkRedisLocationStore =
        SampleLocationStore.create(topology)

    @Bean(destroyMethod = "close")
    fun bingoMetricsReporter(registry: MeterRegistry): BingoMetricsReporter =
        BingoMetricsReporter(registry, "session")

    @Bean(destroyMethod = "close")
    fun bingoReadinessReporter(
        topology: SampleTopology,
        meshes: ZLinkRouteMeshRuntime,
    ): BingoReadinessReporter = BingoReadinessReporter.session(topology, meshes)

    companion object {
        fun run(args: Array<String> = emptyArray()): AutoCloseable {
            val environment =
                StandardEnvironment().apply {
                    propertySources.remove(
                        StandardEnvironment.SYSTEM_ENVIRONMENT_PROPERTY_SOURCE_NAME
                    )
                    propertySources.remove(
                        StandardEnvironment.SYSTEM_PROPERTIES_PROPERTY_SOURCE_NAME
                    )
                }
            val builder =
                SpringApplicationBuilder(SessionServerApplication::class.java)
                    .environment(environment)
                    .properties(
                        "spring.config.location=${Path.of(SampleTopology.configPath(args)).toAbsolutePath().toUri()}"
                    )
                    .web(WebApplicationType.NONE)
            builder.application().setKeepAlive(true)
            val context = builder.run()
            return AutoCloseable { context.close() }
        }
    }
}
